### PR TITLE
TRUNK-6160 Added diagnoses support to encounter service

### PR DESCRIPTION
**Pull request by icrc-loliveira**
## Description of what I changed
Added diagnoses support to encounter service.

## Issue I worked on

see https://issues.openmrs.org/browse/TRUNK-6160

## Checklist: I completed these to help reviewers :)

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/test/java/org/openmrs/api/EncounterServiceTest.java [hunk 1]
```diff
@@ -425,6 +428,33 @@ public void saveEncounter_shouldCascadeSaveToContainedConditions() {
...
+		diagnosis1.setRank(1);
+		diagnosis1.setCertainty(ConditionVerificationStatus.CONFIRMED);
+
+		Diagnosis diagnosis2 = new Diagnosis();
+		diagnosis2.setDiagnosis(new CodedOrFreeText(null, null, "Also fever"));
+		diagnosis2.setRank(2);
+		diagnosis2.setCertainty(ConditionVerificationStatus.PROVISIONAL);
+		
+		// replay
+		Set<Diagnosis> diagnoses = new HashSet<>();
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/api/EncounterServiceTest.java:447]: What are you using this hashset for?

**Comment by dkayiwa** [comment]
@icrc-loliveira your tests do not demonstrate the reported problem because they all pass without a single change in `EncounterServiceImpl`

File: api/src/test/java/org/openmrs/api/EncounterServiceTest.java [hunk 2]
```diff
@@ -1047,6 +1093,23 @@ public void unvoidEncounter_shouldCascadeUnvoidToOrders() {
...
 	}
+
+
+	/**
+	 * @see EncounterService#unvoidEncounter(Encounter)
+	 */
+	@Test
+	public void unvoidEncounter_shouldCascadeUnvoidToDiagnoses() {
+		EncounterService encounterService = Context.getEncounterService();
+
```
Comment (tendomart) [inline api/src/test/java/org/openmrs/api/EncounterServiceTest.java:1104]: Redundant space may be removed

File: api/src/test/java/org/openmrs/api/EncounterServiceTest.java (hunk 1 above)
Comment (icrc-loliveira) [reply api/src/test/java/org/openmrs/api/EncounterServiceTest.java:447]: Nothing, some leftovers from an earlier code.

**Comment by icrc-loliveira** [comment]
Hello @dkayiwa,
I have fixed the test to create diagnoses. (I based my test on the existing ones, so is possible that those are also not validating as expected.)

For the void and unvoided functions. It seems, based on the tests, that they are already taking into account the Diagnoses. So I removed my changes but leave the tests. (I have no idea how this is being handled but if it is working that is good enougt for me :) )

File: api/src/main/java/org/openmrs/api/impl/EncounterServiceImpl.java [hunk 3]
```diff
@@ -447,10 +456,7 @@ public Encounter unvoidEncounter(Encounter encounter) throws APIException {
 			        .getEditPrivilege() });
 		}
 		
-		String voidReason = encounter.getVoidReason();
-		if (voidReason == null) {
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/impl/EncounterServiceImpl.java:451]: Do we really need this change?

File: api/src/test/java/org/openmrs/api/EncounterServiceTest.java [hunk 4]
```diff
@@ -974,7 +1001,23 @@ public void voidEncounter_shouldCascadeToOrders() {
...
+	@Test
+	public void voidEncounter_shouldCascadeVoidToDiagnoses() {
+		EncounterService encounterService = Context.getEncounterService();
+
+		// get a nonvoided encounter that has some obs
+		Encounter encounter = encounterService.getEncounter(1);
+		encounterService.voidEncounter(encounter, "Just Testing");
 
+		Diagnosis diagnosis = Context.getDiagnosisService().getDiagnosis(1); 
+		assertTrue(diagnosis.getVoided());
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/api/EncounterServiceTest.java:1017]: Can you assert that this was not voided before.

File: api/src/test/java/org/openmrs/api/EncounterServiceTest.java [hunk 5]
```diff
@@ -1048,6 +1091,22 @@ public void unvoidEncounter_shouldCascadeUnvoidToOrders() {
...
 	}
 	
+	/**
+	 * @see EncounterService#unvoidEncounter(Encounter)
+	 */
+	@Test
+	public void unvoidEncounter_shouldCascadeUnvoidToDiagnoses() {
+		EncounterService encounterService = Context.getEncounterService();
+
+		// get a voided encounter that has some voided obs
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/api/EncounterServiceTest.java:1101]: Voided obs or voided diagnosis?

File: api/src/test/java/org/openmrs/api/EncounterServiceTest.java [hunk 6]
```diff
@@ -1048,6 +1091,22 @@ public void unvoidEncounter_shouldCascadeUnvoidToOrders() {
...
+		Encounter encounter = encounterService.getEncounter(2);
+		encounterService.unvoidEncounter(encounter);
+
+		Diagnosis diagnosis = Context.getDiagnosisService().getDiagnosis(2);
+		assertFalse(diagnosis.getVoided());
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/api/EncounterServiceTest.java:1106]: Can you assert that this was voided before.

File: api/src/test/java/org/openmrs/api/EncounterServiceTest.java [hunk 7]
```diff
@@ -974,7 +1001,23 @@ public void voidEncounter_shouldCascadeToOrders() {
...
 	}
+	
+	/**
+	 * @see EncounterService#voidEncounter(Encounter,String)
+	 */
+	@Test
+	public void voidEncounter_shouldCascadeVoidToDiagnoses() {
+		EncounterService encounterService = Context.getEncounterService();
+
+		// get a nonvoided encounter that has some obs
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/api/EncounterServiceTest.java:1012]: That has some obs or diagnosis?

File: api/src/test/java/org/openmrs/api/EncounterServiceTest.java (hunk 5 above)
Comment (icrc-loliveira) [reply api/src/test/java/org/openmrs/api/EncounterServiceTest.java:1101]: Done

File: api/src/test/java/org/openmrs/api/EncounterServiceTest.java [hunk 8]
```diff
@@ -1048,6 +1099,22 @@ public void unvoidEncounter_shouldCascadeUnvoidToOrders() {
...
+	 */
+	@Test
+	public void unvoidEncounter_shouldCascadeUnvoidToDiagnoses() {
+		EncounterService encounterService = Context.getEncounterService();
+
+		// get a voided encounter that has some voided Diagnoses
+		Encounter encounter = encounterService.getEncounter(2);
+		encounterService.unvoidEncounter(encounter);
+
+		Diagnosis diagnosis = Context.getDiagnosisService().getDiagnosis(2);
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/api/EncounterServiceTest.java:1113]: Can you assert that the diagnosis was voided before the un-voiding of the encounter?

File: api/src/test/java/org/openmrs/api/EncounterServiceTest.java [hunk 9]
```diff
@@ -1048,6 +1099,22 @@ public void unvoidEncounter_shouldCascadeUnvoidToOrders() {
...
 	
+	/**
+	 * @see EncounterService#unvoidEncounter(Encounter)
+	 */
+	@Test
+	public void unvoidEncounter_shouldCascadeUnvoidToDiagnoses() {
+		EncounterService encounterService = Context.getEncounterService();
+
+		// get a voided encounter that has some voided Diagnoses
+		Encounter encounter = encounterService.getEncounter(2);
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/api/EncounterServiceTest.java:1110]: Can you also assert that the encounter is voided?

File: api/src/test/java/org/openmrs/api/EncounterServiceTest.java [hunk 10]
```diff
@@ -974,7 +1001,31 @@ public void voidEncounter_shouldCascadeToOrders() {
...
+	public void voidEncounter_shouldCascadeVoidToDiagnoses() {
+		EncounterService encounterService = Context.getEncounterService();
+
+		// get a nonvoided encounter that has some Diagnoses
+		Encounter encounter = encounterService.getEncounter(1);
 
+		// Test that diagnoses is unvoided
+		Diagnosis unvoidedDiagnosis = encounter.getDiagnoses().iterator().next();
+		assertEquals(unvoidedDiagnosis.getDiagnosisId(), 1);
+		assertTrue(!unvoidedDiagnosis.getVoided());
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/api/EncounterServiceTest.java:1018]: How about using assertFalse?

File: api/src/test/java/org/openmrs/api/EncounterServiceTest.java [hunk 11]
```diff
@@ -974,7 +1001,31 @@ public void voidEncounter_shouldCascadeToOrders() {
...
+
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/api/EncounterServiceTest.java:1019]: Can you also assert that the encounter is not voided?

**Pull request merged**